### PR TITLE
Post field edit permissions

### PR DESCRIPTION
**Pull request by b0b3rt**
In the process of working on [giving podcasters the ability to add podcast episodes to posts](https://github.com/ForumMagnum/ForumMagnum/pull/5799), realized that we were letting anyone who could pass the `editCheck` on a post the ability to edit a bunch of fields which they likely shouldn've have been allowed to do so.

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1201302964208280/1203083459341156) by [Unito](https://www.unito.io)

### CODE REVIEW

File: packages/lesswrong/lib/collections/posts/collection.ts [hunk 1]
```diff
@@ -68,7 +69,8 @@ makeEditable({
     pingbacks: true,
     permissions: {
       viewableBy: ['guests'],
-      editableBy: ['members', 'sunshineRegiment', 'admins'],
+      // TODO: do we also need to cover userIsPostGroupOrganizer somehow?
```
Comment (jpaddison3) [inline packages/lesswrong/lib/collections/posts/collection.ts:72]: Yes I think you do. I claim you should do this by letting userIsPostGroupOrganizer edit the post metadata and include it in the function.

File: packages/lesswrong/lib/collections/posts/schema.tsx [hunk 2]
```diff
@@ -1179,6 +1179,7 @@ const schema: SchemaType<DbPost> = {
     type: Date,
     control: 'datetime',
     viewableBy: ['guests'],
+    // TODO: Feels like this should be more restrictive
     editableBy: ['members'],
```
Comment (jpaddison3) [inline packages/lesswrong/lib/collections/posts/schema.tsx:1183]: I straightforwardly think it should be. This currently lets anyone promote their post to the frontpage.

Comment (jpaddison3) [inline packages/lesswrong/lib/collections/posts/schema.tsx:1183]:
```suggestion
    editableBy: ['sunshineRegiment', 'admins'],
```

File: packages/lesswrong/lib/collections/posts/collection.ts (hunk 1 above)
Comment (jpaddison3) [reply packages/lesswrong/lib/collections/posts/collection.ts:72]: Reading the functions, currently canUserEditPostMetadata already tries to check if the user is a group organizer by getting it from `group.organizerIds`, whereas `userIsPostGroupOrganizer` checks it by running a query against Localgroups.

**Review by jpaddison3 — APPROVED** [review]
ASTLAC

File: packages/lesswrong/lib/collections/posts/collection.ts (hunk 1 above)
Comment (b0b3rt) [reply packages/lesswrong/lib/collections/posts/collection.ts:72]: I think the issue here is that `group.organizerIds` only exists on posts which were returned via a gql resolver (`groupId` is a foreign key field that adds a `group` resolver).  So on the server side, an organizer might actually not be allowed to edit an event's contents even if they're an organizer, though I'll see if I can figure out how to test that.

Comment (b0b3rt) [reply packages/lesswrong/lib/collections/posts/collection.ts]: It turns out that the current behavior on `master` already doesn't let group co-organizers edit events unless they authored the event post themselves, so this won't be a regression.

**Comment by jimrandomh** [comment]
Unfortunately this breaks the feature where group organizers can edit posts in the group. The issue is that `canUserEditPostMetadata` is written in such a way that it can take either a `DbPost` or a `PostsBase`, but if it gets a `DbPost`, it doesn't have access to the group, so it can't check who the organizers are. And it can't query for the group to make up for the missing join, because (1) it isn't async, and also (2) it doesn't have a ResolverContext (so it can't use loaders, which, for a field-level permission, would mean a truly stupendous quantity of duplicate queries).

So, alas, reverting. One proper fix would be to either extend the field-level permissions system to be able to use async functions and give them a `ResolverContext`. Another fix might be to have the object-level permissions pass on metadata for the field-level permissions to refer to. Either way, the prototype of field-level permission check functions has to change, which is a fairly invasive change.

**Pull request merged**